planner/core/point_get_plan_test.go

```diff
@@ -136,4 +141,13 @@ func (s *testPointGetSuite) TestPointGetPlanCache(c *C) {
 	counter.Write(pb)
 	hit = pb.GetCounter().GetValue()
 	c.Check(hit, Equals, float64(2))
+	tk.MustExec("insert into t (a, b, c) values (18446744073709551615, 4, 4)")
+	tk.MustExec("set @p1=-1")
+	tk.MustExec("set @p2=1")
+	tk.MustExec(`prepare stmt7 from "select a from t where a = ?"`)
+	tk.MustQuery("execute stmt7 using @p1").Check(testkit.Rows())
+	tk.MustQuery("execute stmt7 using @p2").Check(testkit.Rows("1"))
```

Comment:
This test case cannot test `PointGet` plan actually? the plan generated using `@p1` is table_scan + selection, not point get, so `@p2` actually reuses this table_scan + selection plan?

Reply:
@eurekaka oh...... this logic is magic...

> This test case cannot test PointGet plan actually? the plan generated using @p1 is table_scan + selection, not point get, so @p2 actually reuses this table_scan + selection plan?

In my test this can use point-get in two executes, because prepare use `?` instead of `-1` or `1`, so `getNameValuePair` can handle `ParramMarkExpr`, then prepare success a point-get and execute as point-get.

```
mysql> show create table t1;
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                               |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (
  `a` bigint(20) NOT NULL,
  `b` int(11) DEFAULT NULL,
  PRIMARY KEY (`a`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> explain select * from t1 where a = -1;                                                                                                                                                                                                                                                                     +-------------------+-------+------+---------------------------------------------------------+
| id                | count | task | operator info                                           |
+-------------------+-------+------+---------------------------------------------------------+
| TableReader_6     | 1.00  | root | data:TableScan_5                                        |
| └─TableScan_5     | 1.00  | cop  | table:t1, range:[-1,-1], keep order:false, stats:pseudo |
+-------------------+-------+------+---------------------------------------------------------+
2 rows in set (0.01 sec)

mysql> prepare st from "explain select * from t1 where a = ?";
Query OK, 0 rows affected (0.00 sec)

mysql> set @x=-1;
Query OK, 0 rows affected (0.00 sec)

mysql> execute st using @x;                                                                                                                                                                                                                                                                                    
+-------------+-------+------+---------------------+
| id          | count | task | operator info       |
+-------------+-------+------+---------------------+
| Point_Get_1 | 1.00  | root | table:t1, handle:-1 |
+-------------+-------+------+---------------------+
1 row in set (0.00 sec)
```
> same as the first case, but handle column is int64, would the second execute report error?

so for this case:

> same as the first case, but handle column is int64, would the second execute report error?

maybe luck is just fine

> handle column is unsigned, if the first execute is pk = 1, and the second execute is like pk = 18446744073709551615, since we are using int64 as handle type in PointGet, would the second execute cause overflow and wrong result?

will ok for both use point-get plan, but there are another question, we keep param as string datum and try to cast it to int64, so will got a overflow question, it seems not easy to fix so it take another issue in here https://github.com/pingcap/tidb/issues/10111.

:rofl:  it's very stranger -1 cannot use plan cache, and prepare and normal query use different plan..

Reply:
I think the root cause of these problems is that we are using int64 to store the handle in PointGet, while it can be unsigned.